File: TEO/Mav.UI/Mav.Ui/src/app/services/admin/i-assessment.service.ts
import { Observable } from 'rxjs/Observable';
import { GenericResult } from '../../models/generic.result.model';
import { AddAssessmentViewModel } from '../../models/admin/add-assessment.view.model';
import { AssessmentSearchModel } from '../../models/search/assessment.search.model';
import { Paginate } from '../../models/paginated.items';
import { AssessmentListingModel } from '../../models/assessment.listing.model';
import { UpdateAssessmentViewModel } from '../../models/admin/update-assessment.view.model';
import { AddAssessmentEoViewModel } from '../../models/admin/add-assessment-eo.view.model';
import { DuplicateAssessmentViewModel } from '../../models/admin/duplicate-assessment.view.model';

export interface IAssessmentService {

    getAssessmentById<T>(id: number): Observable<T>;

    getAllAssessments(filter: AssessmentSearchModel): Observable<Paginate<AssessmentListingModel>>;

    createAssessment(newItem: AddAssessmentViewModel): Observable<GenericResult<number>>;

    createEoAssessment(newItem: AddAssessmentEoViewModel): Observable<GenericResult<number>>;

    updateAssessment(item: UpdateAssessmentViewModel): Observable<GenericResult<number>>;

    deleteAssessment(id: number): Observable<GenericResult<number>>;

    isEoQualification(assessmentId: number): Observable<boolean>;

    duplicateAssessment(item: DuplicateAssessmentViewModel): Observable<GenericResult<number>>;
}

File: TEO/Mav.Services/Abstract/IAssessmentService.cs
using Mav.Common.Models;
using Mav.Data.Repositories;
using Mav.Models.SearchModels;
using Mav.Models.ViewModels;
using System.Collections.Generic;
using System.Threading.Tasks;

namespace Mav.Services.Abstract
{
    public interface IAssessmentService
    {
        Task<T> GetAssessmentById<T>(int id);

        Task<IPaginate<AssessmentListingModel>> GetAllAssessments(AssessmentSearchModel vm);

        Task<IEnumerable<T>> GetAllAssessmentsEnum<T>(AssessmentSearchModel vm);

        Task<GenericResult<T>> AddAssessment<T>(AddAssessmentViewModel m);

        Task<GenericResult<T>> AddAssessment<T>(AddAssessmentEoViewModel m);

        Task<GenericResult<T>> Update<T>(UpdateAssessmentViewModel m);

        Task<GenericResult<T>> Delete<T>(int id);

        Task<List<T>> GetAssessmentRoles<T>();

        Task<bool> IsEOQualification(int id);

        Task<GenericResult<T>> DuplicateAssessment<T>(DuplicateAssessmentViewModel vm);
    }
}
